modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -123,42 +122,143 @@
     private final Map<String, TableImpl> tables = new ConcurrentHashMap<>();
 
     /** Tables. */
-    private final Map<UUID, TableImpl> tablesById = new ConcurrentHashMap<>();
+    private final Map<IgniteUuid, TableImpl> tablesById = new ConcurrentHashMap<>();
+
+    /**
+     * tableId -> future that adds an ability to await distributed table creation on local node from within the context
+     * of user's table creation intention.
+     *
+     * In other words, awaiting local {@link TableManager#createTableLocally(String, IgniteUuid, List, SchemaDescriptor)}
+     * from within {@link TableManager#createTableAsync(String, Consumer, boolean)}
+     */
+    private final Map<IgniteUuid, CompletableFuture<Table>> createTblIntention = new ConcurrentHashMap<>();
```

Comment:
Seems, we can have only one intention for every single table at any time: either to add, drop or alter, but not any combination.
Maybe we can extend Completable future and enrich it with additional informaion?

Reply:
Possibly the logic with intentions might be reduced.
Because now all places when the intention is applying, an event is also firing (create table intention -> create table event and so on).
It looks like a duplicate behavior, I think possible to use only one of them.

Reply:
Well, seems that local events deserves their own refactoring, reverted so far, as Vlad suggested.